amivapi/auth/sessions.py

```diff
@@ -0,0 +1,204 @@
+                ldap_connector.authenticate_user(username, password)):
+            # Success, sync user and get token
+            updated = ldap_connector.sync_one(username)
+            _prepare_token(item, updated['_id'])
+            app.logger.info(
+                "User '%s' was authenticated with LDAP" % username)
+            return
+
+        # Database, try to find via nethz, mail or objectid
+        users = app.data.driver.db['users']
+        lookup = {'$or': [{'nethz': username}, {'email': username}]}
+        try:
+            objectid = ObjectId(username)
+            lookup['$or'].append({'_id': objectid})
+        except InvalidId:
+            pass  # input can't be used as ObjectId
+        user = users.find_one(lookup)
+
+        if user:
+            app.logger.debug("User found in db.")
+            if verify_password(user, item['password']):
+                app.logger.debug("Login for user '%s' successful." % username)
+                _prepare_token(item, user['_id'])
+                return
+            else:
+                status = "Login failed: Password does not match!"
+                app.logger.debug(status)
+                abort(401, description=debug_error_message(status))
+
+        # Abort if everything else fails
+        status = "Login with db failed: User not found!"
+        app.logger.debug(status)
+        abort(401, description=debug_error_message(status))
+
+
+def _prepare_token(item, user_id):
+    token = b64encode(urandom(256)).decode('utf_8')
+
+    # Make sure token is unique
+    while app.data.find_one("sessions", None, token=token) is not None:
+        token = b64encode(urandom(256)).decode('utf_8')
+
+    # Remove user and password from document
+    del item['username']
+    del item['password']
+
+    # Add token (str) and user_id (ObejctId)
+    item['user'] = user_id
+    item['token'] = token
+
+
+def verify_password(user, plaintext):
+    """Check password of user, rehash if necessary.
+
+    It is possible that the password is None, e.g. if the user is authenticated
+    via LDAP. In this case default to "not verified".
+
+    Args:
+        user (dict): the user in question.
+        plaintext (string): password to check
+
+    Returns:
+        bool: True if password matches. False if it doesn't or if there is no
+            password set and/or provided.
+    """
+    password_context = app.config['PASSWORD_CONTEXT']
+
+    if (plaintext is None) or (user['password'] is None):
+        return False
+
+    is_valid = password_context.verify(plaintext, user['password'])
+
+    if is_valid and password_context.needs_update(user['password']):
+        # update password - hook will handle hashing
+        update = {'password': plaintext}
+        with admin_permissions():
+            patch_internal("users", payload=update, _id=user['_id'])
+    return is_valid
+
+
+# Regular task to clean up expired sessions
+@periodic(datetime.timedelta(days=1))
+def delete_expired_sessions():
```

Comment:
For clarification: as the periodicity is set to 1 per day, if somebody sets a session timeout of 15 mins, it may be that the session only times out after 1 day and 15 mins.
Is this intended?